src/Riok.Mapperly/Descriptors/Mappings/UserMappings/UserImplementedExistingTargetMethodMapping.cs

```diff
@@ -0,0 +1,75 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Riok.Mapperly.Descriptors.Mappings.ExistingTarget;
+using Riok.Mapperly.Emit;
+using Riok.Mapperly.Helpers;
+using Riok.Mapperly.Symbols;
+
+namespace Riok.Mapperly.Descriptors.Mappings.UserMappings;
+
+/// <summary>
+/// Represents a mapping method on the mapper which is implemented by the user.
+/// </summary>
+public class UserImplementedExistingTargetMethodMapping : ExistingTargetMapping, IUserMapping
```

Comment:
I'm not really happy with this existing target method mapping implementing `ITypeMapping` as an existing target mapping can never support `Riok.Mapperly.Descriptors.Mappings.ITypeMapping.Build`.

I quickly poked around a little and I think the cleanest would be to introduce a new `INewInstanceMapping : ITypeMapping` with the build method. Then let the `IExistingTargetMapping` extend `ITypeMapping`.

You can find what I tried here: [latonz:inewinstancemapping](https://github.com/latonz/mapperly/tree/inewinstancemapping) / https://github.com/latonz/mapperly/commit/d3e786f7256d846fdbcfbfe9cf64d13400a11e7d.

What I did (more or less):
* Rename `ITypeMapping` / `TypeMapping` to `INewInstanceMapping` / `NewInstanceMapping`
* Added a new interface `ITypeMapping` and moved the build method from `ITypeMapping` to this interface
* Set the base interface of `INewInstanceMapping` and `IExistingTargetMapping` to `ITypeMapping`
* Adjusted `RuntimeTargetTypeMappingBodyBuilder ` to only consider `INewInstanceMapping`s.
* Adjusted `MappnigCollection`: Separate `Add`, `AddNewInstanceMapping` and `AddExistingTargetMapping` methods.

WDYT about this approach?

Reply:
>  But I think it makes a lot of things easier. The body builders often use the types provided by IMapping.

Yeah I agree. The part I dislike is having to pass down the values/ `ITypeMapping` depending on `IMapping`

I imagined something like the following, where `MappingBuilder` creates an `ITypeMapping` (`INewInstance`) and then creates an object inheriting from `IMapping`, sets `SourceType` and `TargetType` from context and adds `ITypeMapping`.

```C#
interface IMapping
{
    ITypeSymbol SourceType { get; }

    ITypeSymbol TargetType { get; }

    MappingBodyBuildingPriority BodyBuildingPriority { get; }

    ITypeMapping Mapping { get; }
}

tbh the this approach has flaws but it would make it possible to make all the mappings equatable between source generator runs. Not only would this have performance benefits but it would make it easier to do advanced mappings in the future.

When experimenting with reverse mapping, nested ignores/mappings and extra parameters; I would have to modify `MappingCollection` to store mappers based on their type and the `UserImplementedMethodMapping` context. Otherwise, depending on the order of generation, mappers would reuse a `MethodMapping` created by another mapper. Due to my changes the method may ignore or use a parameter that the base `UserImplementedMethodMapping` had not been configured for, resulting in an invalid mapping.

Separating the mappers by context would work but would result in a lot of bloated identical `MethodMappings` being generated. Most of the time these mapping methods would be identical in every way. By making all `ITypeMapping` and `MethodMapping` equatable it would be a lot easier to traverse the "graph" and merge/cull the identical mappers.